include/secp256k1_bulletproofs.h

```diff
@@ -0,0 +1,144 @@
+#ifndef _SECP256K1_BULLETPROOFS_
+# define _SECP256K1_BULLETPROOFS_
+
+# include "secp256k1.h"
+# include "secp256k1_generator.h"
+
+# ifdef __cplusplus
+extern "C" {
+# endif
+
+#include <stdint.h>
+
+/** A function-like macro returning the size, in bytes, of an uncompressed
+ * Bulletproof proving that a value lies in the range [0, 2^(n_bits) - 1]
+ */
+#define SECP256K1_BULLETPROOFS_UNCOMPRESSED_SIZE(n_bits) (194UL + (n_bits) * 64)
+
+/** Maximum size, in bytes, of an uncompressed rangeproof */
+extern const size_t SECP256K1_BULLETPROOFS_RANGEPROOF_UNCOMPRESSED_MAX_LENGTH;
+
+/** The same value, as a C macro so it can be used as C89 array size */
+#define SECP256K1_BULLETPROOFS_RANGEPROOF_UNCOMPRESSED_MAX_LENGTH_ \
+    SECP256K1_BULLETPROOFS_UNCOMPRESSED_SIZE(64)
+
+/** Opaque data structure that holds the current state of an uncompressed
+ * Bulletproof proof generation. This data is not secret and does not need
+ * to be handled carefully, but neither does it have any meaning outside
+ * of the API functions that use it.
+ *
+ * Obviously you should not modify it or else you will get invalid proofs.
+ *
+ * Typical users do not need this structure. If you have more than a few
+ * hundred bytes of memory to spare create a proof in one shot with the
+ * TODO function instead.
+ */
+typedef struct {
+    unsigned char data[160];
+} secp256k1_bulletproofs_prover_context;
+
+/** Opaque structure representing a large number of NUMS generators */
+typedef struct secp256k1_bulletproofs_generators secp256k1_bulletproofs_generators;
+
+/** Allocates and initializes a list of NUMS generators, along with precomputation data
+ *  Returns a list of generators, or NULL if allocation failed.
+ *  Args:          ctx: pointer to a context object (cannot be NULL)
+ *                   n: number of NUMS generators to produce. Should be 128 to allow for
+ *                      64-bit rangeproofs
+ */
+SECP256K1_API SECP256K1_WARN_UNUSED_RESULT secp256k1_bulletproofs_generators *secp256k1_bulletproofs_generators_create(
+    const secp256k1_context* ctx,
+    size_t n
+) SECP256K1_ARG_NONNULL(1);
+
+/** Allocates a list of generators from a static array
+ *  Returns a list of generators, or NULL if allocation or parsing failed.
+ *  Args:      ctx: pointer to a context object (cannot be NULL)
+ *  In:       data: data that came from `secp256k1_bulletproofs_generators_serialize`
+ *        data_len: the length of the `data` buffer
+ */
+SECP256K1_API SECP256K1_WARN_UNUSED_RESULT secp256k1_bulletproofs_generators* secp256k1_bulletproofs_generators_parse(
```

Comment:
Do we actually want parsing and serialization in the public API (especially since I don't see it called anywhere outside of tests)? It seems dangerous to trust a serialized list of generators that we didn't create ourselves (there might be some hidden discrete log relation between them), so we might as well always just call create.

Reply:
> It seems dangerous to trust a serialized list of generators that we didn't create ourselves

Many things in life are dangerous :).

In practice I expect this list to be hardcoded, so yes, we need an API that can just blindly accept them.

Reply:
LOL, fair. Maybe we can include a whitelist of hashes for known safe serialized lists of generators?

Reply:
How long would this list be?

Reply:
Length 1 :) if you want to use something that's not the canonical list of generators then it's on you

Reply:
How many generators are on the list?

Reply:
I was thinking 128

Reply:
128 won't be enough for zkps.